From 622a8922a87d9810cd5bcba3f70a46e10e1c1f12 Mon Sep 17 00:00:00 2001
From: Keith Packard <keithp@keithp.com>
Date: Thu, 13 Feb 2025 12:49:58 -0800
Subject: [PATCH 2/2] rx: allow cmpstrnsi len to be zero

The SCMPU instruction doesn't change the C and Z flags when the
incoming length is zero, which means the insn will produce a
value based upon the existing flag values.

As a quick kludge, adjust these flags to ensure a zero result in this
case.

Signed-off-by: Keith Packard <keithp@keithp.com>
---
 gcc/config/rx/rx.md | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/gcc/config/rx/rx.md b/gcc/config/rx/rx.md
index f5857cb616e..27ba373d9b4 100644
--- a/gcc/config/rx/rx.md
+++ b/gcc/config/rx/rx.md
@@ -2590,7 +2590,9 @@ (define_insn "rx_cmpstrn"
    (clobber (reg:SI 3))
    (clobber (reg:CC CC_REG))]
   "rx_allow_string_insns"
-  "scmpu		; Perform the string comparison
+  "setpsw  z		; Set flags in case len is zero
+   setpsw  c
+   scmpu		; Perform the string comparison
    mov     #-1, %0      ; Set up -1 result (which cannot be created
                         ; by the SC insn)
    bnc	   ?+		; If Carry is not set skip over
-- 
2.47.2

